src/matrixrtc/NewMembershipManager.ts

```diff
@@ -0,0 +1,808 @@
+                            state.sendMembershipRetries++;
+                            this.scheduler.addAction({
+                                ts: Date.now() + Math.max(retryIn, this.callMemberEventRetryDelayMinimum),
+                                type: DelayedLeaveActionType.SendMainDelayedEvent,
+                            });
+                        },
+                        () => {
+                            throw Error(
+                                "Exceeded maximum send delayed event attempts (client._unstable_sendDelayedStateEvent): " +
+                                    e,
+                            );
+                        },
+                    );
+                }
+                break;
+            case DelayedLeaveActionType.SendScheduledDelayedLeaveEvent:
+                if (state.delayId) {
+                    try {
+                        await this.client._unstable_updateDelayedEvent(state.delayId, UpdateDelayedEventAction.Send);
+                        state.hasMemberStateEvent = false;
+                        state.updateDelayedEventRetries = 0;
+                        this.scheduler.resetActions([]);
+                        this.leavePromiseHandle.resolve?.(true);
+                    } catch (e) {
+                        const notFoundHandled = this.handleNotFoundError(e, () => {
+                            state.delayId = undefined;
+                            this.scheduler.addAction({ ts: Date.now(), type: MembershipActionType.SendLeaveEvent });
+                        });
+                        const rateLimitHandled = this.handleRateLimitError(
+                            e,
+                            state.updateDelayedEventRetries,
+                            (retryIn) => {
+                                state.updateDelayedEventRetries++;
+                                this.scheduler.addAction({
+                                    ts: Date.now() + retryIn,
+                                    type: DelayedLeaveActionType.SendScheduledDelayedLeaveEvent,
+                                });
+                            },
+                            () => {
+                                this.scheduler.addAction({ ts: Date.now(), type: MembershipActionType.SendLeaveEvent });
+                            },
+                        );
+                        if (!notFoundHandled && !rateLimitHandled) {
+                            this.scheduler.addAction({ ts: Date.now(), type: MembershipActionType.SendLeaveEvent });
+                        }
+                    }
+                } else {
+                    this.scheduler.addAction({ ts: Date.now(), type: MembershipActionType.SendLeaveEvent });
+                }
+                break;
+        }
+    }
+
+    public async membershipLoopHandler(state: ActionSchedulerState, type: MembershipActionType): Promise<void> {
+        switch (type) {
```

Comment:
I would find it slightly easier to read this if it followed the lifecycle ordering: `SendJoinEvent`, `Update`, `SendLeaveEvent`